crates/rpc/src/v03/method/get_events.rs

```diff
@@ -200,97 +176,222 @@ pub async fn get_events(
+                // we should continue with the first pending event in the next
+                // page
+
+                // check if there are matching pending events
+                let mut buf: Vec<types::EmittedEvent> = Vec::new();
+                let _ =
+                    append_pending_events(&pending_block, &mut buf, 0, 1, request.address, keys);
+                if buf.is_empty() {
+                    // if there are no matching events in pending we should not return a token
+                    events.continuation_token = None;
+                } else {
+                    let continuation_token = ContinuationToken {
+                        block_number: pending_block_number,
+                        offset: 0,
+                    };
+                    events.continuation_token = Some(continuation_token.to_string());
+                }
+            }
+        }
     }
 
-    check_continuation_token_validity(requested_offset, &events.events)?;
+    check_continuation_token_validity(continuation_token, &events.events)?;
 
     Ok(events)
 }
 
+// Handle the case when we're querying events exclusively from the pending block.
+async fn get_pending_events(
+    context: RpcContext,
+    request: &EventFilter,
+    pending_block: Option<Arc<PendingBlock>>,
+    continuation_token: Option<ContinuationToken>,
+) -> Result<types::GetEventsResult, GetEventsError> {
+    match pending_block {
+        None => match continuation_token {
+            None => {
+                // if there was no continuation token we return empty result
+                Ok(types::GetEventsResult {
+                    events: Vec::new(),
+                    continuation_token: None,
+                })
+            }
+            // no pending block means we can't continue from the token offset
+            Some(_) => Err(GetEventsError::InvalidContinuationToken),
+        },
+        Some(pending_block) => {
+            let storage = context.storage.clone();
+
+            let pending_block_parent_hash = pending_block.parent_hash;
+            let pending_block_number = tokio::task::spawn_blocking(move || {
+                let mut connection = storage
+                    .connection()
+                    .context("Opening database connection")?;
+
+                let transaction = connection
+                    .transaction()
+                    .context("Creating database transaction")?;
+
+                let parent_block_header = transaction
+                    .block_header(pathfinder_storage::BlockId::Hash(pending_block_parent_hash))?;
+
+                let pending_block_number = parent_block_header
+                    .map(|header| BlockNumber::new_or_panic(header.number.get() + 1));
+
+                Ok::<_, GetEventsError>(pending_block_number.unwrap_or_default())
+            })
+            .await
+            .context("Fetching block number for pending block")??;
```

Comment:
I'm a little concerned about opening multiple db transactions within the same call. Using the same block hash protects us here; maybe add a comment to that effect so no-one gets any clever ideas in the future.